src/native/pprof.c

```diff
@@ -273,21 +269,67 @@ static uint64_t _pprof_locNew(DProf* dp, uint64_t id_mapping, uint64_t addr, uin
 
 }
 
-uint64_t pprof_locAdd(DProf* dp, uint64_t id_mapping, uint64_t addr, uint64_t* functions, int64_t* lines, size_t n_functions) {
+static bool isEqualLine(const PPLine *line, uint64_t function_id, int64_t lineno) {
+  uint64_t fid_a = line->function_id, fid_b = function_id;
+  int64_t line_a = line->line, line_b = lineno;
+  return !((fid_a ^ fid_b) | (line_a ^ line_b));
+}
+
+static bool isEqualLocation(const PPLocation *location, uint64_t mapping_id,
+                            uint64_t addr, const uint64_t *functions,
+                            const int64_t *lines, size_t n_functions) {
+  if (location->mapping_id != mapping_id) return false;
+  if (location->address != addr) return false;
+  if (location->n_line != n_functions) return false;
+
+  for (size_t i = 0; i < n_functions; ++i) {
+    if (!isEqualLine(location->line[i], functions[i], lines[i])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static PPLocation *findLocation(DProf* dp, uint64_t id_mapping, uint64_t addr,
+                                const uint64_t *functions, const int64_t *lines,
+                                size_t n_functions) {
   PPProfile* pprof = &dp->pprof;
-  for(size_t i=0; i < pprof->n_location; i++) {
-    if(isEqualLocation(id_mapping, addr, pprof->location[i]))
-      return i+1;
+  for (size_t i = 0; i < pprof->n_location; ++i) {
+    PPLocation *location = pprof->location[i];
+    if (isEqualLocation(location, id_mapping, addr, functions, lines, n_functions)) {
+      return location;
```

Comment:
I've historically stayed away from stdbool, so potentially naive question--since C itself has no types smaller than a byte, if I cast the return (after the call returns, _not_ in a debugger on line 301 ors omething) to `(char)` will modern compilers return `1` if the location is nonzero, or will it be some other value (probably a uint64 mod 256 or whatever)?

If the latter, maybe `return !!location`, but no strong feeling either way.

Reply:
I couldn't find a quick reference for conversion _from_ bool to other types in C. It's guaranteed to convert _to_ `0` or `1`, and assume that implication is enough when converting from bool to other types, but I'm not sure.